File: packages/php/tests/WebhookVerifierTest.php
<?php

declare(strict_types=1);

namespace Catalisa\Biometrics\Tests;

use Catalisa\Biometrics\Evidence;
use Catalisa\Biometrics\WebhookVerificationException;
use Catalisa\Biometrics\WebhookVerifier;
use PHPUnit\Framework\TestCase;

/**
 * The vectors are shared by every SDK in this repository and were produced by the
 * building block's own signing code, so "it passes here" means the same thing in
 * each language.
 */
final class WebhookVerifierTest extends TestCase
{
    /** @return array<string,mixed> */
    private function vectors(): array
    {
        return json_decode((string) file_get_contents(__DIR__ . '/../../../vectors/vectors.json'), true);
    }

    private function verifierAt(array $webhook, ?string $when = null, int $tolerance = 300): WebhookVerifier
    {
        return new WebhookVerifier(
            $webhook['keys'],
            $tolerance,
            new \DateTimeImmutable($when ?? $webhook['now']),
        );
    }

    public function testAcceptsAGenuineDelivery(): void
    {
        $webhook = $this->vectors()['webhook'];

        $event = $this->verifierAt($webhook)->verify($webhook['headers'], $webhook['body']);

        self::assertNotEmpty($event['id']);
        self::assertSame('biometrics.session.completed', $event['type']);
    }

    public function testReadsHeadersInPhpServerShape(): void
    {
        // $_SERVER hands them over as HTTP_X_WEBHOOK_ID; a framework as x-webhook-id.
        $webhook = $this->vectors()['webhook'];
        $serverStyle = [];
        foreach ($webhook['headers'] as $name => $value) {
            $serverStyle['HTTP_' . strtoupper(str_replace('-', '_', $name))] = $value;
        }

        $event = $this->verifierAt($webhook)->verify($serverStyle, $webhook['body']);

        self::assertNotEmpty($event['id']);
    }

    public function testRejectsATamperedBody(): void
    {
        $webhook = $this->vectors()['webhook'];

        try {
            $this->verifierAt($webhook)->verify($webhook['headers'], $webhook['tamperedBody']);
            self::fail('a changed byte must fail the signature');
        } catch (WebhookVerificationException $e) {
            self::assertSame(WebhookVerifier::INVALID_SIGNATURE, $e->failure);
        }
    }

    public function testRejectsAnOldDelivery(): void
    {
        // The engine does not refuse old deliveries — the receiver does, or a
        // captured delivery could be replayed forever.
        $webhook = $this->vectors()['webhook'];
        $tenMinutesLater = (new \DateTimeImmutable($webhook['headers']['x-webhook-timestamp']))
            ->modify('+10 minutes')->format(DATE_ATOM);

        try {
            $this->verifierAt($webhook, $tenMinutesLater)->verify($webhook['headers'], $webhook['body']);
            self::fail('outside the tolerance it must be refused');
        } catch (WebhookVerificationException $e) {
            self::assertSame(WebhookVerifier::TIMESTAMP_OUT_OF_TOLERANCE, $e->failure);
        }

        $event = $this->verifierAt($webhook, $tenMinutesLater, 3600)->verify($webhook['headers'], $webhook['body']);
        self::assertNotEmpty($event['id']);
    }

    public function testRejectsMissingHeadersAndUnknownKey(): void
    {
        $webhook = $this->vectors()['webhook'];
        $verifier = $this->verifierAt($webhook);

        try {
            $verifier->verify([], $webhook['body']);
            self::fail('no headers, no verification');
        } catch (WebhookVerificationException $e) {
            self::assertSame(WebhookVerifier::MISSING_HEADERS, $e->failure);
        }

        $unknown = $webhook['headers'];
        $unknown['x-webhook-key-id'] = 'whk_not_mine';
        try {
            $verifier->verify($unknown, $webhook['body']);
            self::fail('an unknown key id must be refused');
        } catch (WebhookVerificationException $e) {
            self::assertSame(WebhookVerifier::UNKNOWN_KEY_ID, $e->failure);
        }

        $oldVersion = $webhook['headers'];
        $oldVersion['x-webhook-signature'] = 'v2=abc';
        try {
            $verifier->verify($oldVersion, $webhook['body']);
            self::fail('another signature version must be refused');
        } catch (WebhookVerificationException $e) {
            self::assertSame(WebhookVerifier::UNSUPPORTED_SIGNATURE_VERSION, $e->failure);
        }
    }

    public function testVerifiesEvidenceOffline(): void
    {
        $evidence = $this->vectors()['evidence'];
        $keys = $evidence['evidenceKeysResponse']['data'];
        $signature = $evidence['evidence']['signature'];

        $out = Evidence::verify($evidence['sessionId'], $evidence['attempt'], $evidence['evidence']['bundleHash'], $signature, $keys);
        self::assertTrue($out['valid']);
        self::assertFalse($out['unknownKey']);

        // Any change to what was signed breaks it — that is the whole point.
        $other = Evidence::verify($evidence['sessionId'], $evidence['attempt'] + 1, $evidence['evidence']['bundleHash'], $signature, $keys);
        self::assertFalse($other['valid']);

        $unknown = Evidence::verify($evidence['sessionId'], $evidence['attempt'], $evidence['evidence']['bundleHash'], ['alg' => 'Ed25519', 'keyId' => 'ev-unknown', 'value' => $signature['value'], 'signedAt' => $signature['signedAt']], $keys);
        self::assertTrue($unknown['unknownKey']);
        self::assertFalse($unknown['valid']);
    }
}

File: packages/java/src/test/java/app/catalisa/biometrics/WebhookVerifierTest.java
package app.catalisa.biometrics;

import com.fasterxml.jackson.databind.JsonNode;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class WebhookVerifierTest {

    private final JsonNode vectors = Vectors.load();

    private JsonNode webhook() {
        return vectors.get("webhook");
    }

    private Map<String, String> headers(String overrideName, String overrideValue) {
        Map<String, String> headers = new HashMap<>();
        webhook().get("headers").fields()
                .forEachRemaining(entry -> headers.put(entry.getKey(), entry.getValue().asText()));
        if (overrideName != null) {
            headers.put(overrideName, overrideValue);
        }
        return headers;
    }

    private static Function<String, String> reader(Map<String, String> headers) {
        return name -> headers.get(name);
    }

    private WebhookVerifier verifier(Instant at, Duration tolerance) {
        Map<String, String> keys = new HashMap<>();
        webhook().get("keys").fields().forEachRemaining(entry -> keys.put(entry.getKey(), entry.getValue().asText()));

        return new WebhookVerifier(keys, tolerance, () -> at);
    }

    private WebhookVerifier verifier() {
        return verifier(Instant.parse(webhook().get("now").asText()), Duration.ofMinutes(5));
    }

    @Test
    void acceptsAGenuineDelivery() {
        var event = verifier().verify(reader(headers(null, null)), webhook().get("body").asText());

        assertFalse(event.id().isEmpty());
        assertEquals(Models.WebhookEvents.SESSION_COMPLETED, event.type());
    }

    @Test
    void rejectsATamperedBody() {
        var error = assertThrows(WebhookVerifier.VerificationException.class,
                () -> verifier().verify(reader(headers(null, null)), webhook().get("tamperedBody").asText()));

        assertEquals(WebhookVerifier.Failure.INVALID_SIGNATURE, error.failure());
    }

    @Test
    void rejectsAnOldDelivery() {
        // The engine does not refuse old deliveries — the receiver does, or a
        // captured delivery could be replayed forever.
        var sentAt = Instant.parse(webhook().get("headers").get("x-webhook-timestamp").asText());
        var late = sentAt.plus(Duration.ofMinutes(10));

        var error = assertThrows(WebhookVerifier.VerificationException.class,
                () -> verifier(late, Duration.ofMinutes(5)).verify(reader(headers(null, null)), webhook().get("body").asText()));
        assertEquals(WebhookVerifier.Failure.TIMESTAMP_OUT_OF_TOLERANCE, error.failure());

        var event = verifier(late, Duration.ofHours(1))
                .verify(reader(headers(null, null)), webhook().get("body").asText());
        assertFalse(event.id().isEmpty());
    }

    @Test
    void rejectsMissingHeadersUnknownKeyAndAnotherVersion() {
        var body = webhook().get("body").asText();

        var bare = assertThrows(WebhookVerifier.VerificationException.class,
                () -> verifier().verify(name -> null, body));
        assertEquals(WebhookVerifier.Failure.MISSING_HEADERS, bare.failure());

        var unknown = assertThrows(WebhookVerifier.VerificationException.class,
                () -> verifier().verify(reader(headers("x-webhook-key-id", "whk_not_mine")), body));
        assertEquals(WebhookVerifier.Failure.UNKNOWN_KEY_ID, unknown.failure());

        var version = assertThrows(WebhookVerifier.VerificationException.class,
                () -> verifier().verify(reader(headers("x-webhook-signature", "v2=abc")), body));
        assertEquals(WebhookVerifier.Failure.UNSUPPORTED_SIGNATURE_VERSION, version.failure());
    }

    @Test
    void verifiesEvidenceOffline() {
        var evidence = vectors.get("evidence");
        List<Models.EvidenceKey> keys = new ArrayList<>();
        evidence.get("evidenceKeysResponse").get("data").forEach(node ->
                keys.add(BiometricsClient.MAPPER.convertValue(node, Models.EvidenceKey.class)));
        var signature = BiometricsClient.MAPPER.convertValue(
                evidence.get("evidence").get("signature"), Models.EvidenceSignature.class);
        var sessionId = evidence.get("sessionId").asText();
        var attempt = evidence.get("attempt").asInt();
        var bundleHash = evidence.get("evidence").get("bundleHash").asText();

        var ok = Evidence.verify(sessionId, attempt, bundleHash, signature, keys);
        assertTrue(ok.valid());
        assertFalse(ok.unknownKey());

        // Any change to what was signed breaks it — that is the whole point.
        assertFalse(Evidence.verify(sessionId, attempt + 1, bundleHash, signature, keys).valid());
        assertFalse(Evidence.verify(sessionId, attempt, "0000", signature, keys).valid());

        var unknownKey = new Models.EvidenceSignature(
                signature.alg(), "ev-unknown", signature.signedAt(), signature.value());
        var unknown = Evidence.verify(sessionId, attempt, bundleHash, unknownKey, keys);
        assertTrue(unknown.unknownKey());
        assertFalse(unknown.valid());
    }
}
